lib/ecto/association.ex

```diff
@@ -978,12 +971,24 @@ defmodule Ecto.Association.ManyToMany do
   end
 
   @doc false
-  def joins_query(%{queryable: queryable, owner: owner,
-                    join_through: join_through, join_keys: join_keys}) do
+  def joins_query(%{owner: owner, join_through_where: join_through_where,
+                    join_through: join_through, join_keys: join_keys} = assoc) do
     [{join_owner_key, owner_key}, {join_related_key, related_key}] = join_keys
+    join_through_query = join_through_query(join_through, join_through_where)
+
     from o in owner,
-      join: j in ^join_through, on: field(j, ^join_owner_key) == field(o, ^owner_key),
-      join: q in ^queryable, on: field(j, ^join_related_key) == field(q, ^related_key)
+      join: j in ^join_through_query, on: field(j, ^join_owner_key) == field(o, ^owner_key),
+      join: q in ^Ecto.Association.combine_assoc_query(assoc, nil), on: field(j, ^join_related_key) == field(q, ^related_key)
+  end
+
+  defp join_through_query(queryable, []), do: queryable
+  defp join_through_query(queryable, nil), do: queryable
+  defp join_through_query(queryable, {module, function, args}) do
+    join_through_query(queryable, apply(Macro.expand(module, __ENV__), function, args))
```

Comment:
We shouldn't need Macro.expand here. We need to evalute the tuple contents at compile time. See the comment on schema.